### PR TITLE
Improved font synchronization

### LINKED ISSUE #230: Load custom fonts in DOMPDF merging with local or vendor fonts
Currently we pass a cache file `dompdf_font_family_cache.php` directly to the `vendor/dompdf/dompdf/lib/fonts` directory and also our custom fonts as well. This is not ideal because every time we need to upgrade DOMPDF we need to copy this.

#### Things we should address:

* Move our custom fonts to a dedicated plugin directory, I suggest: `assets/fonts`
* Copy the custom fonts to the local (uploads folder) fonts
* Load the uploads fonts directory as the `fontDir`

### PR DESCRIPTION
**Pull request by Spreeuw**
This PR introduces a new approach to merging plugin fonts with locally installed fonts, as well as the fonts supplied by dompdf itself. This way, we can use the 'vanilla' dompdf vendor library (instead of adding our own fonts like we used to do), and still make sure that the fonts provided by our plugin are updated in the local fonts folder each time the plugin is updated.

All in all, the following optimizations have been applied:

1. Plugin fonts now live in `assets/fonts`. The mapping to these files is [in the font synchronization class](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/178535e899ea3697256fd5e58ab2a91f19b4c861/includes/class-wcpdf-font-synchronizer.php#L171-L198) (so it doesn't scan that folder automatically), which means that when adding a new font, it should be added to that mapping too.
2. Font paths are always normalized when read, replacing backward (windows-style) slashes with forward slashes, which should prevent paths with mixed directory separators from being parsed incorrectly.
3. Fonts from dompdf are automatically stripped from the local cache file, to make sure it always gets a fresh copy when dompdf is updated. Note that when a custom font is then added, the [`FontMetrics::saveFontFamilies()`](https://github.com/dompdf/dompdf/blob/8768448244967a46d6e67b891d30878e0e15d25c/src/FontMetrics.php#L92-L108) method will add the dompdf supplied fonts to the cache file again, but it will be relative to `$rootDir` - the font files themselves are not copied. This should be fine.
4. This should be compatible with dompdf 0.8.x/1.0.x (array return) as well as 1.1.x (closure style) font cache files.

We should make sure it is compatible with existing installations (current version of the plugin with custom fonts, keeping those custom fonts and loading them correctly after the font synchronization has run), as well as new/clean installations.

I have added a check whether the bundled dompdf version is loaded, so that the Font Synchronizer does not run when another PDF library is utilized: 
https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/e09579a9dc866a93c56393dde4502ff955212946/includes/class-wcpdf-main.php#L689-L693

I have also not tested this with dompdf 1.1.1 yet but it should be compatible with that version too.

closes #230

### CODE REVIEW

**Comment by Spreeuw** [comment]
> I wonder if we should simply skip the copy font routine altogether when the PDF maker class is overridden (using `wpo_wcpdf_pdf_maker` [here](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/75ee093ea9fadda7ff32cd00b96b904a44870d0f/includes/wcpdf-functions.php#L125))

Added in e09579a

**Comment by alexmigf** [comment]
#### Tests

* New installation (DOMPDF 0.8.3)
    * Plugin fonts copied to local folder and cache file created :heavy_check_mark: 
    * Plugin fonts linked to `$fontDir` in cache file :heavy_check_mark: 
    * PDF generated without errors :heavy_check_mark: 
    * Installing custom fonts:
        * Plugin fonts still preset in local folder :heavy_check_mark: 
        * Custom fonts added to local folder :heavy_check_mark: 
        * Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts linked to `$rootDir` :heavy_check_mark: 
        
        
* New installation (DOMPDF 1.0.x)
    * Plugin fonts copied to local folder and cache file created :heavy_check_mark: 
    * Plugin fonts linked to `$fontDir` in cache file :heavy_check_mark: 
    * Installing custom fonts:
        * Plugin fonts still preset in local folder :heavy_check_mark: 
        * Custom fonts added to local folder :heavy_check_mark: 
        * Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts linked to `$rootDir` :heavy_check_mark: 
        
        
* New installation (DOMPDF 1.1.x)
    * Plugin fonts copied to local folder and cache file created :heavy_check_mark: 
    * Plugin fonts linked to `$fontDir` in cache file :heavy_check_mark: 
    * Installing custom fonts:
        * Plugin fonts still preset in local folder :heavy_check_mark: 
        * Custom fonts added to local folder :heavy_check_mark: 
        * Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts linked to `$rootDir`  :heavy_check_mark: 
        
        
* On plugin upgrade with custom fonts (DOMPDF 0.8.3)
    * Plugin fonts stayed in local folder and dist cache file deleted :heavy_check_mark: 
    * Plugin fonts linked to `$fontDir` in cache file :heavy_check_mark: 
    * Installed custom fonts:
        * Custom fonts still present in local folder :heavy_check_mark: 
        * Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts not showing :heavy_check_mark: 
        
        
* On plugin upgrade with custom fonts (DOMPDF 1.0.x)
    * Plugin fonts stayed in local folder and dist cache file deleted :heavy_check_mark: 
    * Plugin fonts linked to `$fontDir` in cache file :heavy_check_mark: 
    * Installed custom fonts:
        * Custom fonts still present in local folder :heavy_check_mark: 
        * Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts not showing :heavy_check_mark: 
        
        
* On plugin upgrade with custom fonts (DOMPDF 1.1.x)
    * Plugin fonts stayed in local folder and dist cache file deleted :heavy_check_mark: 
    * Plugin fonts linked to `$fontDir` in cache file :heavy_check_mark: 
    * Installed custom fonts:
        * Custom fonts still present in local folder :heavy_check_mark: 
        * Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts not showing :heavy_check_mark:

**Review by alexmigf — APPROVED** [review]

**Comment by YordanSoares** [comment]
## ✔️ Tests with dompdf 1.0.2 and 1.1.1

I have tested dompdf 1.0.2 and 1.1.1 in PHP 7.4.1 and 8.0.0 and everything seems to be working well:

- Plugin fonts copied to local folder and cache file created ✔️
- Plugin fonts linked to `$fontDir` in cache file ✔️
- PDF generated without errors ✔️
- Installing custom fonts:
  - Plugin fonts still preset in local folder ✔️
  - Custom fonts added to local folder ✔️
  - Cache file updated with custom fonts and plugin linked to $fontDir, DOMPDF fonts linked to `$rootDir` ✔️

image
_PDF invoice generated using custom fonts with dompdf 1.0.2_

## ⚠️ Tests with dompdf 0.8.3 and 0.8.6

However, I'm having issues with dompdf 0.8.3 and 0.8.6:

### 

### dompdf 0.8.3 (both PHP 7.4.1 and 8.0.0)

I got this fatal error:

```
Fatal error: Class 'Cpdf' not found
#0 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\CanvasFactory.php(57): Dompdf\Adapter\CPDF->__construct('letter', 'portrait', Object(Dompdf\Dompdf))
#1 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\Dompdf.php(299): Dompdf\CanvasFactory::get_instance(Object(Dompdf\Dompdf), 'letter', 'portrait')
#2 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\class-wcpdf-pdf-maker.php(50): Dompdf\Dompdf->__construct(Object(Dompdf\Options))
#3 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\documents\abstract-wcpdf-order-document.php(742): WPO\WC\PDF_Invoices\PDF_Maker->output()
#4 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\documents\abstract-wcpdf-order-document.php(777): WPO\WC\PDF_Invoices\Documents\Order_Document->get_pdf()
#5 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\class-wcpdf-main.php(374): WPO\WC\PDF_Invoices\Documents\Order_Document->output_pdf('inline')
#6 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-includes\class-wp-hook.php(303): WPO\WC\PDF_Invoices\Main->generate_pdf_ajax('')
#7 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-includes\class-wp-hook.php(327): WP_Hook->apply_filters('', Array)
#8 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-includes\plugin.php(470): WP_Hook->do_action(Array)
#9 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-admin\admin-ajax.php(187): do_action('wp_ajax_generat...')
#10 {main}
```

### dompdf 0.8.6 (PHP 7.4.1)

The PDF is generated, but the plugin and custom fonts are not copied to the local folder, nor the references are present in the cache file:

image
_PDF invoice generated using custom fonts with dompdf 0.8.6 and PHP 7.4.1_

### dompdf 0.8.6 (PHP 8.0.0)

I got this fatal error:

```
Fatal error: realpath(): Argument #1 ($path) must be of type string, array given
#0 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\FontMetrics.php(216): realpath(Array)
#1 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\Css\Stylesheet.php(1558): Dompdf\FontMetrics->registerFont(Array, 'C:\\Users\\Yordan...', NULL)
#2 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\Css\Stylesheet.php(1399): Dompdf\Css\Stylesheet->_parse_font_face('font-family: 'g...')
#3 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\Css\Stylesheet.php(332): Dompdf\Css\Stylesheet->_parse_css('@font-face {\r\n\t...')
#4 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\Dompdf.php(659): Dompdf\Css\Stylesheet->load_css('\t\t\t@font-face {...', 3)
#5 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\vendor\dompdf\dompdf\src\Dompdf.php(738): Dompdf\Dompdf->processHtml()
#6 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\class-wcpdf-pdf-maker.php(54): Dompdf\Dompdf->render()
#7 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\documents\abstract-wcpdf-order-document.php(742): WPO\WC\PDF_Invoices\PDF_Maker->output()
#8 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\documents\abstract-wcpdf-order-document.php(777): WPO\WC\PDF_Invoices\Documents\Order_Document->get_pdf()
#9 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips-improved-font-synchronization\includes\class-wcpdf-main.php(374): WPO\WC\PDF_Invoices\Documents\Order_Document->output_pdf('inline')
#10 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-includes\class-wp-hook.php(303): WPO\WC\PDF_Invoices\Main->generate_pdf_ajax('')
#11 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-includes\class-wp-hook.php(327): WP_Hook->apply_filters('', Array)
#12 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-includes\plugin.php(470): WP_Hook->do_action(Array)
#13 C:\Users\Yordan\Local Sites\pdf-invoices-248\app\public\wp-admin\admin-ajax.php(187): do_action('wp_ajax_generat...')
#14 {main}
```

**Comment by Spreeuw** [comment]
@YordanSoares thanks for performing those tests. Can you let me know what versions of PDF invoices you used to do the 0.8.3 & 0.8.6 tests? I think 0.8.3 is not compatible with PHP8.0 (not sure about 7.4)

To be clear, all you would need to do is to install an old version of WCPDF, add a custom font, and _then_ upgrade to the new version (either dompdf 1.0.2 or 1.1.1). I would start with a clean install each time to make sure the situation is representative.

Testing the new class with old versions of dompdf (which I think you may have been trying?) is pointless because nobody will ever have that combination unless they fabricate it themselves.

**Comment by YordanSoares** [comment]
> @YordanSoares thanks for performing those tests. Can you let me know what versions of PDF invoices you used to do the 0.8.3 & 0.8.6 tests? I think 0.8.3 is not compatible with PHP8.0 (not sure about 7.4)

In my first tests, I only used this PR, replacing the dompdf version in every test and deleting the local directory.

> To be clear, all you would need to do is to install an old version of WCPDF, add a custom font, and _then_ upgrade to the new version (either dompdf 1.0.2 or 1.1.1). I would start with a clean install each time to make sure the situation is representative.

Thanks for the explanation, I'll run some extra tests following your instructions.

> Testing the new class with old versions of dompdf (which I think you may have been trying?) is pointless because nobody will ever have that combination unless they fabricate it themselves.

Understood 👍

**Review by YordanSoares — APPROVED** [review]
I run several tests in a clean installation with PHP 7.4.1, updating from WCPDF 2.2.6 (that uses dompdf 0.8.3) to this PR without any issues. Also, I remembered that dompdf library was updated in WCPDF 2.8.0 from 0.8.3 to 1.0.2, precisely because it did not support PHP 8 :)

**Comment by Spreeuw** [comment]
> without any issues.

What did you check?

**Comment by YordanSoares** [comment]
> What did you check?

I checked the following when upgrading from WCPDF v2.2.6 to this PR:

- PDF generated without errors ✔️
- Plugin fonts copied to local folder and cache file created ✔️
- Plugin fonts linked to `$fontDir` in cache file ✔️
- Installing custom fonts:
  - Plugin fonts still present in local folder ✔️
  - Custom fonts added to local folder ✔️
  - Cache file updated with custom fonts and plugin linked to `$fontDir`, DOMPDF fonts linked to `$rootDir` ✔️

**Comment by Spreeuw** [comment]
did you also install custom fonts in 2.2.6 (before the update) and check if they still worked correctly in the new version after updating?

**Comment by YordanSoares** [comment]
@Spreeuw I forgot to answer here: I did all my test using the Custom font extension.
